### PR TITLE
Wake Angle Feature: atan2 direction channel for wake deficit geometry

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #2213 (Wake Deficit Feature, just merged) added gap-normalized fore-TE offsets (dx/gap, dy/gap) as 2 new input channels and delivered a striking **-4.1% p_in** improvement. This establishes that explicitly encoding wake-relative geometry dramatically helps the model understand fore-aft pressure coupling.

However, the current wake features encode *magnitude* components (dx/gap, dy/gap) but not the *direction* of the wake vector as a single canonical angle. Adding a wake angle channel `theta = atan2(dy/gap, dx/gap)` provides:
1. **Angular invariance**: the angle is directly interpretable as the direction of wake impingement independent of gap magnitude
2. **Non-redundant information**: atan2 is NOT derivable from dx/gap and dy/gap without numerical precision issues (it disambiguates the quadrant)
3. **Complement to gap-normalized offsets**: the three features (dx/gap, dy/gap, theta) together form a complete polar coordinate system for each node's wake position

Physical motivation: aerodynamicists parameterize wake interaction by incidence angle. A node at the same angular position from the TE but different distance experiences qualitatively different wake deficit. The wake angle tells the model *from which direction* the fore-foil wake impinges on each aft-foil node.

**Expected improvement:** -1 to -3% additional p_tan, -1 to -2% p_oodc (the one metric wake deficit feature slightly missed).

## Instructions

The `--wake_deficit_feature` flag and implementation are already in the codebase from PR #2213. You need to add one more channel: the wake angle.

### 1. Add argparse flag

```python
parser.add_argument('--wake_angle_feature', action='store_true', default=False,
                    help='Add atan2(dy/gap, dx/gap) wake angle channel alongside wake_deficit_feature')
```

### 2. Implement in `Transolver.forward` (in the wake_deficit_feature block, after computing dx_norm/dy_norm)

In the existing wake deficit block (after `dx_norm` and `dy_norm` are computed), add:

```python
if cfg.wake_deficit_feature and cfg.wake_angle_feature:
    # Compute wake angle (atan2 of gap-normalized offsets)
    # Use raw (un-normalized) offsets for atan2 to preserve angular meaning
    dx_raw = x[:, :, 0] - fore_te_x[:, None]  # [B, N]
    dy_raw = x[:, :, 1] - fore_te_y[:, None]  # [B, N]
    theta = torch.atan2(dy_raw, dx_raw)  # [B, N] in [-pi, pi]
    theta = theta * is_tandem  # zero for single-foil
    wake_features = torch.cat([
        wake_features,  # existing [B, N, 2] from wake_deficit_feature
        theta.unsqueeze(-1)  # [B, N, 1]
    ], dim=-1)  # [B, N, 3]
```

**Note:** This requires `--wake_deficit_feature` to be also set (the wake angle depends on the TE computation from that flag). Raise a clear error or warning if `--wake_angle_feature` is set without `--wake_deficit_feature`.

### 3. Update input dimension

The `wake_deficit_feature` code already does `n_x += 2`. When `wake_angle_feature` is also enabled, add `n_x += 1` after, for a total of `n_x += 3`.

### 4. Run commands (2 seeds, must use BOTH flags)

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/wake-angle-s42" --wandb_group "frieren/wake-angle-feature" \
  --wake_deficit_feature --wake_angle_feature \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame

# Seed 73: same with --seed 73 --wandb_name "frieren/wake-angle-s73"
```

**Note:** `--te_coord_frame` is included — it is in the baseline. The wake features re-use the fore-TE computation from `te_coord_frame` (refactored in PR #2213).

## Baseline

**Current best (PR #2213 — Wake Deficit Feature, 2-seed avg):**

| Metric | Value | Merge threshold |
|--------|-------|-----------------|
| p_in | **11.979** | < 11.98 |
| p_oodc | **7.643** | < 7.65 |
| p_tan | **28.341** | < 28.34 |
| p_re | **6.300** | < 6.30 |

W&B runs: hgml7i2r (seed 42), qic03vrg (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Wake Angle Feature (2-seed avg)

| Metric | Baseline (Wake Deficit) | + Wake Angle | Δ | Merge threshold |
|--------|------------------------|-------------|---|-----------------|
| **p_tan** | 28.341 | 28.685 | **+1.2%** ❌ | < 28.34 |
| p_in | 11.979 | 12.602 | **+5.2%** ❌ | < 11.98 |
| p_oodc | 7.643 | 7.798 | +2.0% ❌ | < 7.65 |
| p_re | 6.300 | 6.417 | +1.9% ❌ | < 6.30 |

### Per-seed breakdown (mae_surf_p)

| Seed | p_tan | p_in | p_oodc | p_re | W&B |
|------|-------|------|--------|------|-----|
| 42 | 29.656 | 13.037 | 7.551 | 6.482 | sa1xdfb7 |
| 73 | 27.714 | 12.166 | 8.045 | 6.351 | xvdyqpd7 |

### Commands
```bash
cd cfd_tandemfoil && python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/wake-angle-s42" --wandb_group "frieren/wake-angle-feature" \
  --wake_deficit_feature --wake_angle_feature --te_coord_frame \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### Peak memory
~46.3-46.8 GB (same as baseline)

### Training duration
~146-147 epochs in 181 min (same throughput)

### What happened

**Wake angle hurts all metrics — negative result.** Adding atan2(dy_raw, dx_raw) as a 3rd wake channel regressed performance across all splits.

1. **All metrics regressed**: p_tan +1.2%, p_in +5.2%, p_oodc +2.0%, p_re +1.9%. The angle channel appears to add noise rather than useful signal. Fails all merge thresholds.

2. **High inter-seed variance**: p_tan ranges from 27.71 (s73) to 29.66 (s42) — a 2.0 spread vs baseline's 0.78 spread. The wake angle may be causing optimization instability.

3. **Why it likely failed**: The atan2 value is *redundant* with the existing (dx/gap, dy/gap) channels — the angle IS derivable from the Cartesian components. While numerically atan2 resolves quadrant ambiguity, the MLP preprocessing layer can learn this mapping. Adding the raw angle creates a confounding third feature that the model must reconcile with the existing two, likely increasing the effective input noise.

4. **Contrast with wake deficit's success**: The wake deficit feature (dx/gap, dy/gap) succeeded because it provided genuinely new information (gap normalization + TE-relative positioning). The wake angle adds only a redundant encoding of the same information — the model can already compute angular relationships from the Cartesian offsets.

5. **Scale issue**: atan2 produces values in [-π, π] while dx/gap and dy/gap can range widely. This scale mismatch may cause the model to over-attend to the angle channel early in training, disrupting the learned representations.

### Suggested follow-ups

- **This idea should be dropped.** The wake deficit feature already captures the relevant information in a more model-friendly form (Cartesian coordinates).
- **If angular encoding is still desired**, use sin/cos decomposition (sin(theta), cos(theta)) instead of raw atan2, which avoids the discontinuity at ±π and normalizes the scale. But given the redundancy analysis, this is likely still unhelpful.
- **Alternative**: focus on *non-redundant* wake features — e.g., wake width (chord-normalized), local turbulence intensity proxy, or Reynolds-scaled distance.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2221 because All 4 metrics regressed (p_tan +1.2%, p_in +5.2%, p_oodc +2.0%, p_re +1.9%). Your analysis is exactly right: atan2(dy, dx) is mathematically derivable from the existing (dx/gap, dy/gap) channels — the MLP preprocessing can learn this mapping, so the wake angle is redundant information, not new information. The scale mismatch ([-π, π] vs varying Cartesian scales) and discontinuity at ±π likely inject noise rather than signal. High seed variance (p_tan 29.66 vs 27.71 = 1.95 gap vs baseline's 0.78) confirms instability. The wake deficit feature succeeded because it added genuinely new geometry-invariant information (gap normalization). Future wake features should follow the same principle: add information the model cannot derive from existing channels. Wake angle direction exhausted..
